File: src/data/make_dataset.py
"""
Pipeline de Ingeniería de Datos (ETL) y Privacidad para TheLook.
Flujo: Raw Data -> Cleaning -> Dynamic Cutoff (Anti-Leakage) -> Feature Engineering -> Anonymization.
Malla de seguridad: Microsoft Presidio (Scan-Mode).
"""
import pandas as pd
import numpy as np
import os
from src.features.anonymizer import PIIAnonymizer
from datetime import datetime

def make_dataset(raw_dir, processed_dir):
    """
    Transforma los datos crudos en el dataset maestro de Churn.
    Implementa el filtrado temporal estricto (Shift-Left) para prevenir Data Leakage
    basado en una ventana de inactividad de 120 días (Standard CLV Window).
    """
    print("🧪 Iniciando procesamiento de datos...")
    
    # Verificamos que los archivos descargados existan
    required_files = ["orders.csv", "order_items.csv", "events.csv", "users.csv", "products.csv"]
    for f in required_files:
        if not os.path.exists(os.path.join(raw_dir, f)):
            raise FileNotFoundError(f"No se encontró {f} en {raw_dir}")

    # 1. Carga de datos crudos
    print("📖 Cargando archivos CSV...")
    orders = pd.read_csv(os.path.join(raw_dir, "orders.csv"))
    order_items = pd.read_csv(os.path.join(raw_dir, "order_items.csv"))
    events = pd.read_csv(os.path.join(raw_dir, "events.csv"))
    users = pd.read_csv(os.path.join(raw_dir, "users.csv"))
    products = pd.read_csv(os.path.join(raw_dir, "products.csv"))

    # Convertir fechas
    order_items['created_at'] = pd.to_datetime(order_items['created_at'], format='mixed', utc=True)
    events['created_at'] = pd.to_datetime(events['created_at'], format='mixed', utc=True)

    # 2. Definir Fechas Clave (Evitar Leakage)
    MAX_DATE = order_items['created_at'].max()
    CHURN_WINDOW = 120
    CUTOFF_DATE = MAX_DATE - pd.Timedelta(days=CHURN_WINDOW)
    
    print(f"📅 Fecha Máxima: {MAX_DATE}")
    print(f"📅 Fecha de Corte (Features): {CUTOFF_DATE}")
    print(f"📅 Ventana de Target Churn: {CHURN_WINDOW} días")

    # 3. FILTRADO AGUAS ARRIBA (Upstream Filtering)
    # Solo lo que pasó antes o en el corte es material para FEATURES
    valid_status = ['Complete', 'Shipped', 'Processing']
    
    order_items_clean = order_items[
        (order_items['status'].isin(valid_status)) & 
        (order_items['created_at'] <= CUTOFF_DATE)
    ].copy()

    returns_clean = order_items[
        (order_items['status'] == 'Returned') & 
        (order_items['created_at'] <= CUTOFF_DATE)
    ].copy()

    events_clean = events[events['created_at'] <= CUTOFF_DATE].copy()

    # 4. Agregación RFM (Basada exclusivamente en datos CLEAN)
    print("🧮 Calculando métricas RFM (Honestas)...")
    order_level = order_items_clean.groupby(['user_id', 'order_id']).agg(
        order_revenue=('sale_price', 'sum')
    ).reset_index()
    
    aov_per_user = order_level.groupby('user_id').agg(
        avg_order_value=('order_revenue', 'mean')
    ).reset_index()

    user_features = order_items_clean.groupby('user_id').agg(
        total_orders=('order_id', 'nunique'),
        total_items=('id', 'count'),
        total_revenue=('sale_price', 'sum'),
        first_purchase=('created_at', 'min'),
        last_purchase=('created_at', 'max'),
        unique_products=('product_id', 'nunique')
    ).reset_index()

    user_features = user_features.merge(aov_per_user, on='user_id', how='left')

    # Recency y Tenure relativos al CUTOFF_DATE
    user_features['recency_days'] = (CUTOFF_DATE - user_features['last_purchase']).dt.days
    user_features['customer_tenure_days'] = (CUTOFF_DATE - user_features['first_purchase']).dt.days
    
    # Malla de Seguridad (Diagnóstico)
    future_leakage_count = (user_features['recency_days'] < 0).sum()
    if future_leakage_count > 0:
        print(f"⚠️ ¡ALERTA CRÍTICA!: Se detectaron {future_leakage_count} usuarios con leakage. Revisar upstream.")
    
    user_features['purchase_span_days'] = (user_features['last_purchase'] - user_features['first_purchase']).dt.days
    user_features['avg_days_between'] = np.where(
        user_features['total_orders'] > 1,
        user_features['purchase_span_days'] / (user_features['total_orders'] - 1),
        0
    )

    # 5. Devoluciones (Basadas en returns_clean)
    returns_agg = returns_clean.groupby('user_id').size().reset_index(name='return_count')
    user_features = user_features.merge(returns_agg, on='user_id', how='left').fillna({'return_count': 0})
    user_features['return_rate'] = user_features['return_count'] / user_features['total_items']

    # 6. Eventos Web (Basados en events_clean)
    print("🌐 Procesando eventos web (Honestos)...")
    event_pivot = events_clean.groupby(['user_id', 'event_type']).size().unstack(fill_value=0).reset_index()
    event_pivot.columns = [f"events_{c}" if c != 'user_id' else c for c in event_pivot.columns]
    
    user_features = user_features.merge(event_pivot, on='user_id', how='left').fillna(0)

    # 7. Demografía
    user_demo = users[['id', 'age', 'gender', 'country', 'traffic_source']].rename(
        columns={'id': 'user_id', 'traffic_source': 'signup_source'}
    )
    user_features = user_features.merge(user_demo, on='user_id', how='left')

    # 8. Definición de Target (Basada en la ventana futura)
    # Un usuario es Churn si NO compró nada en el periodo (CUTOFF_DATE, MAX_DATE]
    print("🎯 Definiendo Target Churn (Ventana Futura)...")
    users_with_purchases_post_cutoff = order_items[
        (order_items['status'].isin(valid_status)) & 
        (order_items['created_at'] > CUTOFF_DATE) &
        (order_items['created_at'] <= MAX_DATE)
    ]['user_id'].unique()
    
    user_features['is_churned'] = (~user_features['user_id'].isin(users_with_purchases_post_cutoff)).astype(int)

    # 9. Sanitización PII Industrial (Opción B)
    from dotenv import load_dotenv
    import hashlib
    load_dotenv()
    
    SALT = os.getenv("USER_SALT", "default_salt_antigravity")
    
    def salt_hash(user_id):
        hash_input = f"{user_id}{SALT}".encode()
        return hashlib.sha256(hash_input).hexdigest()[:12]

    print("🔒 Aplicando seudonimización con Salt...")
    original_unique_ids = user_features['user_id'].nunique()
    
    # Aplicar hash y renombrar
    user_features['Internal_ID'] = user_features['user_id'].apply(salt_hash)
    
    # Validación de Colisiones (Crítico para 12 chars)
    hashed_unique_ids = user_features['Internal_ID'].nunique()
    if original_unique_ids != hashed_unique_ids:
        print(f"❌ ¡ERROR DE COLISIÓN!: Se perdieron {original_unique_ids - hashed_unique_ids} IDs en el hashing.")
        # Aquí se podría aumentar la longitud del hash si fuera necesario
    else:
        print(f"✅ Seudonimización exitosa sin colisiones ({hashed_unique_ids} IDs únicos).")

    # Eliminar ID original
    user_features = user_features.drop(columns=['user_id'])

    # --- SHIFT-LEFT PRIVACY: Auditoría de PII Directa en ETL ---
    print("🛡️ Iniciando Auditoría de Privacidad con IA (Presidio)...")
    anonymizer = PIIAnonymizer()
    
    # Columnas técnicas que sabemos que son seguras o ya seudonimizadas
    WHITELIST = [
        'Internal_ID', 'age', 'gender', 'country', 'signup_source', 
        'total_orders', 'total_items', 'total_revenue', 'avg_order_value',
        'recency_days', 'customer_tenure_days', 'purchase_span_days', 'avg_days_between',
        'return_count', 'return_rate', 'is_churned'
    ]
    
    # Escaneamos el dataframe completo
    pii_report = anonymizer.scan_dataframe(user_features, sample_size=50, ignore_cols=WHITELIST)
    
    if pii_report:
        print(f"⚠️ PII DETECTADA en columnas: {list(pii_report.keys())}")
        print("🛠️ Aplicando anonimización batch proactiva...")
        user_features = anonymizer.anonymize_dataframe(user_features, pii_report)
    else:
        print("✅ No se detectó PII adicional en las columnas seleccionadas.")

    # Guardar resultado principal con timestamp
    os.makedirs(processed_dir, exist_ok=True)
    timestamp = datetime.now().strftime("%Y%m%d")
    output_path = os.path.join(processed_dir, f"user_features_churn_{timestamp}.csv")
    user_features.to_csv(output_path, index=False)
    print(f"✅ Dataset seudonimizado guardado en: {output_path}")

    # 10. Saneamiento del 'Huérfano de Eventos' (Opción B)
    event_features_path = os.path.join(processed_dir, "user_event_features.csv")
    if os.path.exists(event_features_path):
        print("🔒 Sanitizando archivo de eventos (user_event_features.csv)...")
        df_ev = pd.read_csv(event_features_path)
        if 'user_id' in df_ev.columns:
            df_ev['Internal_ID'] = df_ev['user_id'].apply(salt_hash)
            df_ev = df_ev.drop(columns=['user_id'])
            df_ev.to_csv(event_features_path, index=False)
            print(f"✅ Archivo de eventos seudonimizado con éxito.")
        else:
            print("⚠️ El archivo de eventos ya parece estar sanitizado o no tiene user_id.")
    else:
        print("⚠️ No se encontró el archivo user_event_features.csv para sanitizar.")

if __name__ == "__main__":
    make_dataset("data/raw", "data/processed")

File: src/data/make_dataset.R
# ==============================================================================
# 🚀 TheLook Churn: Pipeline de Procesamiento de Datos (R Version)
# ==============================================================================
# Este script replica fielmente la lógica de make_dataset.py para R Studio.
# Garantiza: Integridad Temporal y Privacidad (Zero Trust).
# ==============================================================================

# 1. Carga de Librerías (Equivalentes a Pandas/OS)
if (!require(\"dplyr\")) install.packages(\"dplyr\")
if (!require(\"lubridate\")) install.packages(\"lubridate\")
if (!require(\"readr\")) install.packages(\"readr\")
if (!require(\"digest\")) install.packages(\"digest\") # Para Hashing SHA-256

library(dplyr)
library(lubridate)
library(readr)
library(digest)

# 2. Configuración de Rutas
RAW_DIR <- \"data/raw\"
PROCESSED_DIR <- \"data/processed\"
SALT <- \"default_salt_antigravity\" # Debe coincidir con Python para trazabilidad

# 3. Carga de Datos
message(\"🧪 Cargando archivos CSV crudos...\")
orders <- read_csv(file.path(RAW_DIR, \"orders.csv\"), show_col_types = FALSE)
order_items <- read_csv(file.path(RAW_DIR, \"order_items.csv\"), show_col_types = FALSE)
events <- read_csv(file.path(RAW_DIR, \"events.csv\"), show_col_types = FALSE)
users <- read_csv(file.path(RAW_DIR, \"users.csv\"), show_col_types = FALSE)

# 4. Manejo de Fechas y Definición de CUTOFF (Anti-Leakage)
order_items <- order_items %>%
  mutate(created_at = as_datetime(created_at))

MAX_DATE <- max(order_items$created_at, na.rm = TRUE)
CHURN_WINDOW <- days(120)
CUTOFF_DATE <- MAX_DATE - CHURN_WINDOW

message(paste(\"📅 Fecha Máxima:\", MAX_DATE))
message(paste(\"📅 Fecha de Corte (Honesta):\", CUTOFF_DATE))

# 5. Filtrado Aguas Arriba (Upstream Filtering)
valid_status <- c('Complete', 'Shipped', 'Processing')

order_items_clean <- order_items %>%
  filter(status %in% valid_status, created_at <= CUTOFF_DATE)

events_clean <- events %>%
  mutate(created_at = as_datetime(created_at)) %>%
  filter(created_at <= CUTOFF_DATE)

# 6. Agregación RFM y Features de Usuario
message(\"🧮 Calculando métricas RFM...\")

user_features <- order_items_clean %>%
  group_by(user_id) %>%
  summarise(
    total_orders = n_distinct(order_id),
    total_items = n(),
    total_revenue = sum(sale_price),
    first_purchase = min(created_at),
    last_purchase = max(created_at)
  ) %>%
  mutate(
    recency_days = as.integer(as.duration(last_purchase %--% CUTOFF_DATE) / ddays(1)),
    customer_tenure_days = as.integer(as.duration(first_purchase %--% CUTOFF_DATE) / ddays(1))
  )

# 7. Eventos Web (Pivot simple)
message(\"🌐 Procesando eventos web...\")
event_features <- events_clean %>%
  group_by(user_id, event_type) %>%
  summarise(n = n(), .groups = 'drop') %>%
  tidyr::pivot_wider(names_from = event_type, values_from = n, names_prefix = \"events_\", values_fill = 0)

user_features <- user_features %>%
  left_join(event_features, by = \"user_id\") %>%
  mutate(across(starts_with(\"events_\"), ~replace_na(., 0)))

# 8. Definición de Target Churn (Ventana Futura)
users_with_purchases_post_cutoff <- order_items %>%
  filter(status %in% valid_status, created_at > CUTOFF_DATE) %>%
  pull(user_id) %>%
  unique()

user_features <- user_features %>%
  mutate(is_churned = if_else(user_id %in% users_with_purchases_post_cutoff, 0, 1))

# 9. Anonimización (Zero Trust Hashing)
# Aplicar el mismo SHA-256 + Salt que en Python
message(\"🔒 Aplicando seudonimización SHA-256 con Salt...\")

salt_hash <- function(id, salt) {
  # serialize=FALSE asegura que se comporte como un string raw igual que Python
  sapply(id, function(x) substr(digest(paste0(x, salt), algo=\"sha256\", serialize=FALSE), 1, 12))
}

user_features <- user_features %>%
  mutate(Internal_ID = salt_hash(user_id, SALT)) %>%
  select(-user_id) # Destruir ID original

# 10. Guardar Resultado
if (!dir.exists(PROCESSED_DIR)) dir.create(PROCESSED_DIR, recursive = TRUE)
output_path <- file.path(PROCESSED_DIR, \"user_features_churn_R.csv\")
write_csv(user_features, output_path)

message(paste(\"✅ Dataset procesado en R guardado en:\", output_path))
